### DISCUSSION TITLE
Resolved: Website

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#1628] [General]
I'd like to propose we create a website www.rust-bitcoin.org containing, at first:
- Intro blurb
- Basic code snippet
- links to
  - cookbook
  - github/rust-bitcoin
  - github rust-bitcoin discussions
  - irc
  - list of projects using rust-bitcoin crate?
- crate graph?

The main motivation is so we have somewhere to host the cookbook.

I'm useless at webdev, so unless someone else wants to do it, I'll do it using https://gohugo.io (probably with the [sepctral](https://themes.gohugo.io/themes/spectral/) theme) and host in on github pages and use lets encrypt for TLS. Hugo uses markdown for content so terminal-maxis can edit the site.

The whole thing will be open source and I'll just hack up a basic skeleton and hope that the community comes together and makes it better.

I've registered the domain already.

If anyone enjoys devops or webdev I'm totally happy to not be the one setting it up :)

**Comment by tcharding** [comment]
If github pages is acceptable then @apoelstra if you set up a hello wold repo https://pages.github.com/ I can take it from there (should only take about 5 minutes to do).

**Reply by apoelstra** [reply]
Sure: https://github.com/rust-bitcoin/rust-bitcoin.github.io/

I have added you with "maintain" privileges. If anyone else woud like access please ping me here. I think while we're just playing it's fine to push to master or to do "open a PR and merge it yourself without acks" style development, though eventually we'll tighten that up.

**Reply by tcharding** [reply]
Cool, cheers.

**Reply by Kixunil** [reply]
I'm interested, will try to add contents to the cookbook once the skeleton is ready.

**Comment by tcharding** [comment]
This is the outline for the cookbook: https://github.com/rust-bitcoin/rust-bitcoin/issues/1463#issuecomment-1349699022

A quick exploration of the nomicon shows it uses `mdbook` and looks simple enough that we could imitate it.

**Comment by sanket1729** [comment]
strong concept ACK.

**Comment by Harshil-Jani** [comment]
Hey guys !! I am new to rust and bitcoin both ecosystems. I spot this as a good and introductory way for me to get into it. Can I take some of the tasks from the list as a beginner and try to help here along with learning them?

**Reply by apoelstra** [reply]
That would be very welcome! Right now we use the "Good first issue" label on issues to identify tasks. This discussion is about the website...we should probably have something on the website for newbies, but we probably can't have an up-to-date list of specific tasks since it would come out of sync too early.

Is there anything in particular you'd like to see on the website that would be helpful for you as a newbie?

**Reply by Harshil-Jani** [reply]
Maybe !! I can help with the cookbook. I have had a `mdbook` written in past so I can set the minimalistic version for the cookbook if no one has yet took up the initiative and from their we can progressively keep picking things. Works ?

**Reply by apoelstra** [reply]
cc @tcharding would this be helpful? I think you're made more progress toward starting the book than any of us

**Reply by tcharding** [reply]
All I've done so far is look at another mdbook page and assumed we could work it out. Would be awesome to get some help @Harshil-Jani! You could PR to this repo and I'll work out how to get it online.

**Reply by Harshil-Jani** [reply]
Sure !! I would love to take this up. Thanks.

**Comment by tcharding** [comment]
Hi team,

I've hacked up a website using a repository on my own github account (that can and should be move to this org when everyone is happy with it).  In the [source code repo](https://github.com/tcharding/www.rust-bitcoin.org) there is a `build.sh` script that pushes to `master` branch of https://github.com/rust-bitcoin/rust-bitcoin.github.io which is currently being served at https://rust-bitcoin.github.io/ I didn't bother doing PRs because I was lazy and just wanted to see it up. Web dev is not my thing so the idea is to just get something up then others can PR to improve it.

For some reason the cookbook that @Harshil-Jani put together is not being served, it is working when I run `hugo serve` locally so you can do that to see the whole thing. I'll have another go soon, that's enough web dev for now.

Soon as I get an ok from y'all I'll point rust-bitcoin.org and www.rust-bitcoin.org at at. Once the repo moves to rust-bitcoin org we can add a github action to run the deploy script.

Big thanks to @Harshil-Jani for kicking off the cookbook.

**Reply by Harshil-Jani** [reply]
> For some reason the cookbook that @Harshil-Jani put together is not being served, it is working when I run hugo serve locally so you can do that to see the whole thing. I'll have another go soon, that's enough web dev for now.

@tcharding This seems to occur because, I have added `book/` in `.gitignore` so locally, You might be having a `book/` directory filled up with generated HTML contents. But after you push your code, git would simply ignore it. This is why it is not being served after a push. I have added it because I thought we  would dump HTMLs to another branch and keep only markdowns on main or master. But it seems with the website implementation you did, we don't need it anymore. You can delete gitignore and push the `book/` directory.

**Reply by tcharding** [reply]
BOOM, well done lad! It was as you describe, fixed now. Cookbook is officially online :)

**Comment by apoelstra** [comment]
This is really amazing guys. Thanks so much @Harshil-Jani and @tcharding !

**Reply by tcharding** [reply]
Shall we move the www.rust-bitcoin.org repo over to rust-bitcoin then? Is everyone ok with the setup, one repo for the source code and one repo for the site built from the source. And also is everyone ok with using github actions to call `build.sh --deploy` to deploy the built site to master branch of `https://rust-bitcoin.github.io/`?

**Reply by apoelstra** [reply]
That all sounds good to me, though I'm not sure if rust-bitcoin.org ~is a real domain name that we own~ you said just above that you were grabbing it, awesome.

**Reply by Kixunil** [reply]
If it was possible to have one repo it'd decrease the noise at the GH front page but if not it's OK.

**Reply by tcharding** [reply]
We can have a single repo but then the website source code is tied to github pages by way of the repo name, if we ever want to host it somewhere else (*cough* microsoft) then I think its better to have the soure in its own repo. Said another way, the current set up splits the source code and the  built hosted code since the hosted code has to be in a repo with the name dictated by github but we may one day want to host it someplace else.

If that is all overly complicated we can do the branch dance that @Harshil-Jani suggested, but that comes with its own complexities. I like the separation.

**Reply by tcharding** [reply]
Yes I registered the domain already, if there is some public open entity I am happy to transfer the domain to it so its not owned by me personally.

**Reply by apoelstra** [reply]
No, if it weren't you personally I'd suggest me personally :).

**Comment by tcharding** [comment]
I showed the site to my son, his response was, "its a bit plain" :)

**Reply by Kixunil** [reply]
It feels to me like having some slogan would help. E.g. "Build Bitcoin applications with confidence!"

**Reply by apoelstra** [reply]
Good idea, and I like your suggestion.

Over time I'd like to see us use the blog to soapbox a bit about software supply chains -- talk about why we have our dependency policy or MSRV policy, talk about how we use crev (and how others can get started too!) once we start doing that, talk about building a deterministic compiler toolchain using guix (I think we could convince a guix expert to contribute such a thing once we get some more momentum), etc. Then we may want to rethink our slogal to be broader.

**Reply by tcharding** [reply]
My thoughts exactly, a soap box all of our own :)

**Comment by tcharding** [comment]
Hey @apoelstra can you please edit the settings in rust-bitcoin.github.io to add a custom domain (steps 1 - 4 https://docs.github.com/en/pages/configuring-a-custom-domain-for-your-github-pages-site/managing-a-custom-domain-for-your-github-pages-site#configuring-an-apex-domain). Please use rust-bitcoin.org  I've configured www subdomain using CNAME

I only just set the DNS so might need to propagate still.

**Reply by apoelstra** [reply]
Done. Heh, from my end it looks like it's working. Cheers.

**Comment by tcharding** [comment]
Hey @Harshil-Jani, I've created a new repository here under this organisation to hold the website https://github.com/rust-bitcoin/www.rust-bitcoin.org, while doing so I just pushed all the commits for the repo at `tcharding/www.rust-bitcoin.org`. In doing so I realize now that there is no trail in the git index to indicate your work setting up the cookbook skeleton. My apologies for that. You still have your name listed in the authors section so that is some attribution at least. I just wanted to make you aware that it was not done intentionally. Thanks.

**Reply by Harshil-Jani** [reply]
No worres !! Thanks for informing.

**Comment by yancyribbens** [comment]
This is pretty cool.  It would be nice if the blog supported rss so folks can subscribe.

**Comment by tcharding** [comment]
Good idea, thanks. https://github.com/rust-bitcoin/www.rust-bitcoin.org/issues/7

**Comment by tcharding** [comment]
Website exists, closing.
